### PR TITLE
 src/settings.vala: Introduce 'high-contrast-icon-theme-name' gsetting and toggle icon theme if a11y indicator's high contrast switch gets toggled.

### PR DESCRIPTION
**Pull request by sunweaver**

### CODE REVIEW

File: src/arctica-greeter.vala [hunk 1]
```diff
@@ -1071,12 +1071,24 @@ public class ArcticaGreeter : Object
         /* Set GTK+ settings */
         debug ("Setting GTK+ settings");
         var settings = Gtk.Settings.get_default ();
-        var value = AGSettings.get_string (AGSettings.KEY_THEME_NAME);
+        var agsettings_ = new AGSettings ();
```
Comment (Ionic) [inline src/arctica-greeter.vala]: I'd NAK this.

There was a reason why I have implemented it the way it currently is: starting AG in normal mode, we can set `gtk-theme-name` and `gtk-icon-theme-name` at the start to the values provided via gsettings and THEN create the first `AGSetting` object, which will query the "default" GTK (icon) theme (which is what we read from gsettings OR the fallback that GTK went to if the data we set was wrong) and cache it for us. This is how I handled a "faulty" theme name in gsettings, if I remember correctly. 

You're breaking this logic with this change: you're now constructing an `AGSettings` object without having set the (icon) theme name previously, so `AGSettings` will query the actual default GTK (icon) theme, which differs from what we'd like to know.

Fortunately, we don't need any of this, because [`ArcticaGreeter`'s `ready_cb` function](https://github.com/ArcticaProject/arctica-greeter/pull/70/commits/abd6d8d598b7ceb051976439e8527f549716d96d#diff-0de1eca16ff24dff9fb4f820051f18ca30fabcb13a42044d931f1aa5eca8e42dL429) will synchronize the high contrast and big font values on startup, which means that users shouldn't notice anything changing.

Actually testing this again, though, it looks like my original idea doesn't work out too well: if the theme name we provide via `gtk-theme-name` is invalid, GTK will fall back to a different theme (probably the one called "Default" or Adwaita), but sadly not update the `gtk-theme-name` property. I guess we can get rid of the default-theme-name stuff in `src/settings.vala` instead and settle on a hybrid of what we had before and what you tried to do: drop this commit and change the other ones, like I'll explain in different comments.

File: src/settings.vala [hunk 2]
```diff
@@ -164,20 +164,31 @@ public class AGSettings : Object
...
+                debug ("Switching icon theme to high contrast theme \"%s\"", icon_theme_name);
             }
             else
             {
-                debug ("Switching GTK Theme to default theme \"%s\"", this.default_theme_name_);
-                debug ("Switching icon theme to default icon theme \"%s\"", this.default_icon_theme_name_);
-                settings.set ("gtk-theme-name", this.default_theme_name_);
-                settings.set ("gtk-icon-theme-name", this.default_icon_theme_name_);
+
+                /* FIXME: We need to check for wrong theme names here) and handle such flaws gracefully */
```
Comment (Ionic) [inline src/settings.vala]: This FIXME is difficult to resolve. It's difficult to query GTK themes - there exists no API for doing so, so we'd have to look at file system content in `/usr/share/themes/*/gtk-{3,4}.0` and any potentially user-provided location (which might not be an issue for AG, since the lightdm user is a system user and shouldn't have custom themes).

File: src/settings.vala [hunk 3]
```diff
@@ -164,20 +164,31 @@ public class AGSettings : Object
...
+
+                theme_name = AGSettings.get_string (AGSettings.KEY_THEME_NAME);
+                icon_theme_name = AGSettings.get_string (AGSettings.KEY_ICON_THEME_NAME);
+                debug ("Switching GTK Theme to default theme \"%s\"", theme_name);
+                debug ("Switching icon theme to default icon theme \"%s\"", icon_theme_name);
             }
+            settings.set ("gtk-theme-name", theme_name);
+            settings.set ("gtk-icon-theme-name", icon_theme_name);
         }
     }
```
Comment (Ionic) [inline src/settings.vala:201]: Remove `default_theme_name_` here.

File: src/settings.vala [hunk 4]
```diff
@@ -198,4 +201,5 @@ public class AGSettings : Object
     private bool high_contrast_ = AGSettings.get_boolean (AGSettings.KEY_HIGH_CONTRAST);
     private bool big_font_ = AGSettings.get_boolean (AGSettings.KEY_BIG_FONT);
     private string default_theme_name_;
+    private string default_icon_theme_name_;
```
Comment (Ionic) [inline src/settings.vala]: Don't introduce `default_icon_theme_name_` at all, given that we'd only drop it later.

File: src/settings.vala [hunk 5]
```diff
@@ -163,17 +166,17 @@ public class AGSettings : Object
...
+                settings.set ("gtk-icon-theme-name", AGSettings.get_string (AGSettings.KEY_HIGH_CONTRAST_ICON_THEME_NAME));
             }
             else
             {
-                /*
                 debug ("Switching GTK Theme to default theme \"%s\"", this.default_theme_name_);
-                */
+                debug ("Switching icon theme to default icon theme \"%s\"", this.default_icon_theme_name_);
                 settings.set ("gtk-theme-name", this.default_theme_name_);
+                settings.set ("gtk-icon-theme-name", this.default_icon_theme_name_);
```
Comment (Ionic) [inline src/settings.vala]: `AGSettings.get_string (AGSettings.KEY_ICON_THEME_NAME)` instead.

**Review by Ionic — CHANGES_REQUESTED** [review]
I'd like to rework this, as explained in the in-line comments.

File: src/settings.vala (hunk 5 above)
Comment (sunweaver) [reply src/settings.vala]: yes, see follow-up change

File: src/settings.vala (hunk 3 above)
Comment (sunweaver) [reply src/settings.vala:201]: Maybe... I wonder if we shouldn't use default_theme_name_ (resolving to Adwaita) as a fallback, in case someone puts a typo into KEY_THEME_NAME, similarly for default_icon_theme_name and KEY_ICON_THEME_NAME.

For highcontrast-mode, we could either hard-core "HighContrast" (and depend our packages on gnome-accessibility-themes) or also use Adwaita as fallback.

File: src/settings.vala (hunk 4 above)
Comment (sunweaver) [reply src/settings.vala]: Here, as well: I have in mind using the default_(icon_)theme_name_ variables to resolve the FIXME of missing theme fallbacks

File: src/arctica-greeter.vala (hunk 1 above)
Comment (sunweaver) [reply src/arctica-greeter.vala]: I don't want the greeter to start in some Gtk default theme or the normal theme and then switch over to high-contrast theming immediately after coming up, because AGSettings say so.

In my tests with this branch, I observed AG coming up in normal theme mode and then blend over into high-contrast mode. This causes to much visual noise for visual impaired users, imho.

Also, setting Adwaita first (gtk default theme) and then the normal theme does not make sense here, either. As we obviously have access to all those static values in AGSettings, I don't see any problem using them before AGSettings is fully initialized.

What we can avoid is the pre-mature initialization of "agsettings_". We could also do a static call like`AGSettings.get_boolean(AGSettings.KEY_HIGH_CONTRAST)`.

Basically, the only change needed here is dropping "agsettings_ initialization and usage. IMHO.

File: src/settings.vala (hunk 2 above)
Comment (sunweaver) [reply src/settings.vala]: How about try / catch? I assume (haven't tested, yet) that Gtk will throw an error if we try to set a theme / icon theme that does not exist locally. If so, we simply could catch that and fall-back to Gtk's default theme / icon theme.

**Review by sunweaver — COMMENTED** [review]
I am not convinced about changing this MR heavily.

Also, due to time constraint, please leave the FIXMEs for now. It is more important to get the release rolling with a11y features in place.

If you are concerned about the FIXMEs, please file a reminder bug. Thanks.

File: src/settings.vala (hunk 2 above)
Comment (Ionic) [reply src/settings.vala]: Unfortunately, no, that won't work. The property is just copied directly in GTK (which can't fail) and failure to load a theme is not reported in any way.

File: src/arctica-greeter.vala (hunk 1 above)
Comment (Ionic) [reply src/arctica-greeter.vala]: Okay, that's fair. I concur.

In that case, though, I'd like to move the agsettings object creation to the place where you use it first, including the comment I've written for it.

Additionally, we should drop all the private `default_*theme` variables, it doesn't make sense to keep them. Querying the default themes won't help us. They are hardcoded in GTK, loaded up by default and loading a different theme will just silently do nothing if it's not available. We can't really detect that. We also don't care, because if the strings are invalid, nothing will happen and GTK will stay on its default.

File: src/settings.vala (hunk 3 above)
Comment (Ionic) [reply src/settings.vala:201]: I've looked up GTK's source code and it just doesn't make sense to make things more complicated than necessary here.

Instead, we should really just drop the `default_*_` variables and not query them in construct at all.

If invalid strings are provided, GTK will just do nothing and stay on the theme it currently has loaded.

In high contrast mode, we should certainly try to detect an invalid theme and just hardcode `HighContrast`, but there's no way that we could get this information from GTK. This makes our `default_*_` variables completely useless.

File: src/settings.vala (hunk 5 above)
Comment (Ionic) [reply src/settings.vala]: Did you mean to (force) push some change? I haven't seen any follow-up change.

File: src/arctica-greeter.vala (hunk 1 above)
Comment (sunweaver) [reply src/arctica-greeter.vala]: ok, go ahead then with this... Thanks!

File: src/settings.vala (hunk 2 above)
Comment (sunweaver) [reply src/settings.vala]: Oh dear, but this also means that a flawed configuration will result in Adwaita and nothing more. So, Gtk handles the defaulting. Fair enough, so we can drop that part from our code. Actually a good solution as it simplifies our code.

File: src/settings.vala (hunk 3 above)
Comment (sunweaver) [reply src/settings.vala:201]: Yes, here I concur. Go ahead and rip those `default_*_*` private properties out again.

File: src/settings.vala (hunk 5 above)
Comment (sunweaver) [reply src/settings.vala]: Ignore, we drop the `default_*_` properties and we are done here.

**Comment by Ionic** [comment]
Rebased branch with the proposed changes, please take a look.

I can get rid of the FIXMEs and do actual error handling at a later time. I pretty much already know how to.

**Comment by sunweaver** [comment]
> Rebased branch with the proposed changes, please take a look.
> 
> I can get rid of the FIXMEs and do actual error handling at a later time. I pretty much already know how to.

LGTM now. Thanks!

**Comment by sunweaver** [comment]
> Rebased branch with the proposed changes, please take a look.
> 
> I can get rid of the FIXMEs and do actual error handling at a later time. I pretty much already know how to.

Would it make sense to work on this rather sooner than later (not that you forget what you have in mind here and now)?
